### DISCUSSION TITLE
Cannot get TFT_eSPI to work with KALINCO P22 Smart Watch

### DISCUSSION THREAD

**Discussion by ljunquera** [Bodmer/TFT_eSPI#2377] [Compatible displays and setup files]
I am really struggling to get this working with a simple demo with a ST7789, which I saw in a [YouTube video](https://youtu.be/WFVjsxFMbSM?t=228). My screen gets backlit but nothing else happens. Here is the code I'm looking to run:

```
#include </Users/lj/Projects/TFT_eSPI/TFT_eSPI.h>
#include <SPI.h>

TFT_eSPI tft = TFT_eSPI();

void setup() {
  // put your setup code here, to run once:
  tft.init();
  tft.fillScreen(TFT_RED);
}

void loop() {
  // put your main code here, to run repeatedly:
}
```

I downloaded the repo to a local directory to [remove of some legacy code](https://github.com/Bodmer/TFT_eSPI/discussions/2371) in the ST7789_init.h. I am using a custom configuration in /Projects/TFT_eSPI_Setups/my_custom_setup.h according to the [tips](https://github.com/Bodmer/TFT_eSPI#tips). I have made the change in my User_Setup_Select.h file:

`#include <../TFT_eSPI_Setups/my_custom_setup.h>`

in my_custom_setup I have the settings I need:

```
// ST7789 240 x 240 display with no chip select line
#define USER_SETUP_ID 24

#define ST7789_DRIVER     // Configure all registers

#define TFT_WIDTH  240
#define TFT_HEIGHT 240

#define TOUCH_CS -1

//#define TFT_RGB_ORDER TFT_RGB  // Colour order Red-Green-Blue
//#define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red

//#define TFT_INVERSION_ON
//#define TFT_INVERSION_OFF


//my edits:
#define TFT_MISO 4
#define TFT_MOSI 3
#define TFT_SCLK 2
#define TFT_CS   -1      // Define as not used
#define TFT_BL	 23
#define TFT_BACKLIGHT_ON LOW
#define TFT_DC 18
#define TFT_RST 13
```

I took the pin settings from code that worked with the device with a different codebase (not TFT_eSPI):
```
//display
#define LCD_SCK      2
#define LCD_SDI      3
#define LCD_CS       25
#define LCD_RESET    26
#define LCD_RS       18  //this actually looks to be the DC because I see where it is sending a command
#define LCD_DET      9

//Flash Chip
#define SPI_SCK      2     //mapped to TFT_SCLK 
#define SPI_MOSI     3     //mapped to TFT_MOSI
#define SPI_MISO     4     //mapped to TFT_MISO
#define SPI_CE       5     // UNMAPPED

//Touchscreen
#define TP_SDA       6     // UNMAPPED
#define TP_SCL       7     // UNMAPPED


#define TP_RESET     13     //mapped to TFT_RST
#define TP_INT       28     // maybe TFT_DC

//Backlight
#define LCD_BACKLIGHT_LOW  14     // mapping high only
#define LCD_BACKLIGHT_MID  22     // mapping high only
#define LCD_BACKLIGHT_HIGH 23     //mapped to TFT_BL
```

**Comment by Bodmer** [comment]
If the display worked with the other library and those pins then the TFT_eSPI setup should be:

#define TFT_MISO 4
#define TFT_MOSI 3
#define TFT_SCLK 2
#define TFT_CS   25
#define TFT_BL	 23
#define TFT_BACKLIGHT_ON LOW
#define TFT_DC 18
#define TFT_RST 26

If that does not work, run Read_User_Setup and look at the serial output to check the compiler is picking up the pins correctly.  If that looks OK then provide information on the processor you are using and a link to the display vendors web page so I can see the technical details.

**Reply by ljunquera** [reply]
Thank you! Will do. I need to get the serial output running.

**Comment by ljunquera** [comment]
I'm doing something wrong. Why are some of the settings there but most are GPIO -1?
```
[code]
TFT_eSPI ver = 2.5.21
Processor    = Generic
Transactions = Yes
Interface    = SPI
Display driver = 7789
Display width  = 240
Display height = 240
MOSI    = GPIO -1
MISO    = GPIO -1
SCK     = GPIO -1
TFT_DC   = GPIO -1
TFT_RST  = GPIO -1
TFT_BL           = GPIO -1
TFT_BACKLIGHT_ON = LOW
Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled
Display SPI frequency = 40.00
[/code]
```

My User_Setup_Select.h has these lines in it (everything else is commented except the "do not tinker" section):
```
#ifndef USER_SETUP_LOADED //  Lets PlatformIO users define settings in

#include <../TFT_eSPI_Smartwatch_Setups/kalinco_p22_setup.h>
//#include <User_Setup.h>           // Default setup is root library folder

#endif // USER_SETUP_LOADED
```

my kalinco_p22_setup.h has this:

```
#define ST7789_DRIVER     // Configure all registers

#define TFT_WIDTH  240
#define TFT_HEIGHT 240

#define TOUCH_CS -1

//#define TFT_RGB_ORDER TFT_RGB  // Colour order Red-Green-Blue
//#define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red

//#define TFT_INVERSION_ON
//#define TFT_INVERSION_OFF

//my edits:
#define TFT_MISO 4
#define TFT_MOSI 3
#define TFT_SCLK 2
#define TFT_CS   25      // Define as not used
#define TFT_BL   23
#define TFT_BACKLIGHT_ON LOW
#define TFT_DC   18
#define TFT_RST  26
//#define TFT_RST  13
```


[Here's the code](https://github.com/Bodmer/TFT_eSPI/blob/5886b2cb13751670d2a4e5c0c5eb042c851b1d08/TFT_eSPI.h#L68) in the code which picks up the User_Setup_Select.h. So either, I'm linking it wrong, which would cause an error I presume, or something is overwriting my values.

**Reply by Bodmer** [reply]
What processor and display are you using?  Links to product webpages would be helpful.

**Reply by Bodmer** [reply]
Incidentally you can add things like:
```
#define USER_SETUP_ID 108
constexpr char setup_file[] =  "USER_SETUP_ID 108";
```
to your setup file, then you can see which setup file has been used if you put this in the sketch:
```
void setup()
{
  Serial.begin(115200);
  while(!Serial);
  tft.init();
  tft.fillScreen(TFT_BLACK);
  Serial.println(setup_file);
  Serial.println(USER_SETUP_ID);
 .... etc
```

**Reply by ljunquera** [reply]
If I'm setting `#define TFT_MISO 4 ` in a file and the Read_User_Setup.ino is showing `MISO    = GPIO -1` does that mean it's a setup problem vs a hardware problem?

I see a line during compilation:

```
Using cached library dependencies for file: /Users/lj/Documents/Arduino/libraries/TFT_eSPI/TFT_eSPI.cpp
```

When I have this in the Read_User_Setup.ino:

```
#include </Users/lj/Projects/tft/TFT_eSPI/TFT_eSPI.h>
```

Why would that be?

**Reply by ljunquera** [reply]
I changed the include to:

`#include </Users/lj/Projects/tft/TFT_eSPI/TFT_eSPI.h>`

and it still seems to use the cached one. I did see this in the Arduino Output when compiling:

```
Alternatives for User_Setup_Select.h: [TFT_eSPI@2.5.0]
ResolveLibrary(User_Setup_Select.h)
  -> candidates: [TFT_eSPI@2.5.0]
...
Using cached library dependencies for file: /Users/lj/Library/Arduino15/packages/smartwatch-boards/hardware/nRF5/0.1.6/libraries/SPI/SPI.cpp
Using cached library dependencies for file: /Users/lj/Documents/Arduino/libraries/TFT_eSPI/TFT_eSPI.cpp
```

The Alternatives for User_Setup_Select.h, what does that mean? It can't find the one I want? It says version 2.5.0, which is not my local one.

**Comment by Bodmer** [comment]
It looks like you have multiple copies of the TFT_eSPI library installed! Probably best to delete ALL copies and start again.  Then load the TFT_eSPI library from the Arduino_IDE menu "manage libraries" option to make sure it goes in the correct libraries folder. Edit only the User_Setup.h file with your settings to keep things simple at this stage. Don't try to change anything else in the library!

**Reply by Bodmer** [reply]
Ah, just spotted you are using an nrf5 processor. I assumed from the YouTube tutorial link you gave that you were using an ESP32 on a TTGO board.

The Read_User_Setup sketch will need to be changed as the NRF5 will be treated as a generic processor. [Edit this line](https://github.com/Bodmer/TFT_eSPI/blob/master/examples/Test%20and%20diagnostics/Read_User_Setup/Read_User_Setup.ino#L187) to read:
```
return pin;
```

Then run the sketch again.

**Reply by Bodmer** [reply]
This looks like the board package you are using. 
https://github.com/micooke/arduino-nRF5-smartwatches
In the examples the smart watch has acompletely different display fitted so I need to see the product details. Post a link to the sellers page for the smart watch you are trying to use. I will then looks to see if the technical aspects are compatible with TFT_eSPI.

**Comment by ljunquera** [comment]
Changing line 187, like you said gave me this:
```
[code]
TFT_eSPI ver = 2.5.0
Processor    = Generic
Transactions = Yes
Interface    = SPI
Display driver = 7789
Display width  = 240
Display height = 240

MOSI    = GPIO 3
MISO    = GPIO 4
SCK     = GPIO 2
TFT_CS   = GPIO 25
TFT_DC   = GPIO 18
TFT_RST  = GPIO 26

TFT_BL           = GPIO 23
TFT_BACKLIGHT_ON = LOW

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

Display SPI frequency = 40.00
[/code]
```

The product information is [here](https://manuals.plus/kalinco/p22-smart-watch-manual). If you need more I can do some digging. I actually have a variation of [ATCwatch ](https://github.com/atc1441/ATCwatch) running on it, so I have the settings, thought I may not be sure exactly where they are.

I'm trying to do something really simple to start, like:
`  tft.fillScreen(TFT_RED);`

**Reply by Bodmer** [reply]
It does look like it has a ST7789 screen so it might work with TFT_eSPI however I can't spend any more time on this as it is outside my domain of knowledge.

**Comment by ljunquera** [comment]
Ok. Thanks for the help. I does have a ST7789. Looking through the code on the ATCwatch, I believe I have the TFT_CS, TFT_DC, and TFT_RST settings correct (25, 18, and 26 respectively). Do you have any last advice on where I should focus to try and solve this?

**Reply by Bodmer** [reply]
Maybe the display must be powered on via a control signal?

**Comment by ljunquera** [comment]
The backlight is coming on when I upload, does that mean it's powered on?

**Reply by Bodmer** [reply]
The light and display driver chip might be on separate switches but that is just a guess.
